NBitcoin/Transaction.cs

```diff
@@ -1591,13 +1588,12 @@ public virtual void ReadWrite(BitcoinStream stream)
 			}
 			else
 			{
-				var version = (witSupported && (vin.Count == 0 && vout.Count > 0)) ? nVersion | NoDummyInput : nVersion;
-				stream.ReadWrite(ref version);
+				stream.ReadWrite(ref nVersion);
 
 				if (witSupported)
 				{
 					/* Check whether witnesses need to be serialized. */
-					if (HasWitness)
+					if (HasWitness || Inputs.Count == 0)

```

Comment:
I don't think you need to gate setting the flag value on if there're inputs or not. Instead, if writing a witness encoding is requested and _any_ of the inputs has a witness, then the market byte should be written.

Reply:
From BIP 144: 
> If the witness is empty, the old serialization format must be used.

https://github.com/bitcoin/bips/blob/master/bip-0144.mediawiki#user-content-Serialization:~:text=If%20the%20witness%20is%20empty%2C%20the%20old%20serialization%20format%20must%20be%20used.

Reply:
bitcoind also has a check for failing parsing if the witness encoding is used, but no witnesses are found in the transaction.

Reply:
Ah gotcha, maybe it's used for serializing PSBTs where the template is known but inputs haven't been added? 

I think you should still be able to tighten up this parsing, while still allowing for writing and reading transactions that actually have no inputs.

Reply:
After further analysis, it was only used in tests, and in a single place to estimate the size of a transaction... So I will just forbid to serialize a tx without any input from now.